File: 678_Valid_Parenthesis_String.cpp
#include<bits/stdc++.h>

using namespace std;

class Solution {
private:
    bool checkParenthesis(string s, int i, int cnt){
        if (cnt < 0) return false;
        for (int j = i; j < s.length(); ++j){
            if (s[j] == '(') cnt++;
            else if (s[j] == ')'){
                if (cnt <= 0) return false;
                cnt--;
            }
            else {
                if (checkParenthesis(s, j + 1, cnt)) return true;
                if (checkParenthesis(s, j + 1, cnt + 1)) return true;
                if (checkParenthesis(s, j + 1, cnt - 1)) return true;
            }
        }
        return cnt == 0;
    }
public:
    bool checkValidString2(string s) {
        /* O(3^num(*)) : TLE */
        if (s.empty()) return true;
        return checkParenthesis(s, 0, 0);
    }

    bool checkValidString(string s){
        /* O(n) : 100% */
        stack<int> left, star;
        for (int i = 0; i < s.length(); ++i){
            if (s[i] == '(') left.push(i);
            else if (s[i] == '*') star.push(i);
            else {
                if (left.empty() && star.empty()) return false;
                else if (!left.empty()) left.pop();
                else star.pop();
            }
        }
        while (!left.empty() && !star.empty()){
            if (left.top() > star.top()) return false;
            left.pop(); star.pop();
        }
        return left.empty();
    }
};

int main(){
    Solution* solution = new Solution();
    cout << solution->checkValidString("((()))()(())(*()()())**(())()()()()((*()*))((*()*)"
) << endl;
    return 0;
}

File: 678_Valid_Parenthesis_String.py
class Solution(object):    
    def dfs(self, cnt, s):
        if cnt < 0: 
            return False
        for j, c in enumerate(s):
            if c == '*':
                return self.dfs(cnt - 1, s[j + 1:]) or \
                       self.dfs(cnt + 1, s[j + 1:]) or \
                       self.dfs(cnt, s[j + 1:])
            elif c == '(': 
                cnt += 1
            elif c == ')':
                if cnt <= 0: 
                    return False
                cnt -= 1
            
        return cnt == 0

    def checkValidString(self, s):
        """
        :type s: str
        :rtype: bool
        """
        left, star = [], []
        for i, c in enumerate(s):
            if c == '*':
                star += [i]
            elif c == '(':
                left += [i]
            else:
                if left:
                    left.pop()
                elif star:
                    star.pop()
                else:
                    return False
        while left and star:
            if left[-1] > star[-1]:
                return False
            else:
                left.pop()
                star.pop()
        return len(left) == 0